### PR TITLE
Add/test relationship

### PR DESCRIPTION
**Pull request by jusi-dev**
Created in main branch the new resources to abstract phoney from web-stuff.

PR includes:
- Tests to test every new resource individually
- Test relationship to set / remove favorited contacts from a user

### CODE REVIEW

File: test/accounts/create_user_test.exs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  use PhoneyApi.DataCase
+
+  alias PhoneyApi.Accounts.User
+
+  test "create user" do
+
+    assert {:ok, user} = User.create(%{username: "user1", email: "mail@test.com", password: "password"})
+    assert user.username == "user1"
+    assert user.email == "mail@test.com"
+    assert user.password == "password"
```
Comment (abdelaz3r) [inline test/accounts/create_user_test.exs:11]: Very good!

File: test/contacts/add_favorite_test.exs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    assert {:ok, user} = User.create(%{username: "user1", email: "mail@test.com", password: "password"})
+    assert {:ok, contact} = Contact.create(%{fullname: "contact1", email: "mycontact@test.com", phone_number: "123456789", address: "1234 test street"})
+
+    assert {:ok, user} = User.toggle_favorite(user, contact.id)
+
+    assert user.username == "user1"
+    assert Enum.any?(user.favorites, fn favorite -> favorite.id == contact.id end)
+
+    assert {:ok, user} = User.toggle_favorite(user, contact.id)
+    assert Enum.empty?(user.favorites)
```
Comment (abdelaz3r) [inline test/contacts/add_favorite_test.exs:18]: Very very good.

Personally I find it more readable sometime, instead of using `Enum` functions (ie. `Enum.find`, `Enum.any?`, ...) to juste use pattern matching.

Example:
```elixir
assert user.username == "user1"
# we assert that there is only one favorite
assert [favorite] = user.favorites 
# we assert that the favorite is the correct one
assert favorite.id == contact.id
```

But this is very much personnal preferences, and what you did is 100% correct ;)

File: lib/phoney_api/accounts/user.ex [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  actions do
+    defaults [:read, :destroy, create: :*, update: :*]
+
+    update :favorite_contact do
+      require_atomic? false
+      argument :contact_id, :uuid
+
+      change manage_relationship(:contact_id, :favorites, on_lookup: :relate, on_no_match: :error, on_match: :destroy, on_missing: :create)
+    end
```
Comment (abdelaz3r) [inline lib/phoney_api/accounts/user.ex:39]: That's good (and way less code code that before hehe).

However, and that's the tricky part of `manage_relationship` I think I can make your code fail with the following test:

```elixir
test "add multiple favorites" do
  {:ok, user} =
    User.create(%{
      username: "user1",
      email: "mail@test.com",
      password: "password"
    })

  {:ok, contact1} =
    Contact.create(%{
      fullname: "contact1",
      email: "mycontact@test.com",
      phone_number: "123456789",
      address: "1234 test street"
    })

  {:ok, contact2} =
    Contact.create(%{
      fullname: "contact2",
      email: "mycontact2@test.com",
      phone_number: "123456789",
      address: "1234 test street"
    })

  assert {:ok, user} = User.toggle_favorite(user, contact1.id)
  assert [favorite] = user.favorites
  assert favorite.id == contact1.id

  assert {:ok, user} = User.toggle_favorite(user, contact2.id)
  assert [favorite1, favorite2] = user.favorites
  assert favorite1.id == contact1.id
  assert favorite2.id == contact2.id

  assert {:ok, user} = User.toggle_favorite(user, contact1.id)
  assert [favorite] = user.favorites
  assert favorite.id == contact2.id
end
```

**Review by abdelaz3r — CHANGES_REQUESTED** [review]
Very good!

However you didn't completly get the `manage_relationship` (and that's okay). You can add the test I mentionned on my comment earlier and try to fix your action.

I hope the test is self-explanatory, otherwise don't hesitate to ask me anything on slack.

I think it would be good if you added a few more tests to see if every edge-cases are covered ;)

**Review by abdelaz3r — APPROVED** [review]
